// mini-router/lib/index.mjs
// mini-router — small Request → Response router with middleware.
//
// Shape mirrors itty-router / hono / sunder production routers.
// Real-world consumers: Cloudflare Workers, Bun.serve handlers,
// any HTTP-on-fetch backend.
//
// Features:
//   - HTTP method dispatch (GET/POST/PUT/DELETE/PATCH/OPTIONS)
//   - Path param extraction: /users/:id → req.params.id
//   - Wildcard route: "*" matches all
//   - Middleware chain: handler signature (req, ctx) → Response|undefined
//     If handler returns undefined, chain continues; first Response wins.
//   - Global middleware via .use(handler) — runs before any route
//   - Async handlers throughout (Promise-aware)
//   - Built-in JSON helper (Router.json(data, init?))
//
// Out of scope: regex route patterns, route mounting, error middleware
// chain (single thrown error → 500).

const METHODS = ["GET", "POST", "PUT", "DELETE", "PATCH", "OPTIONS", "HEAD"];

function pathToRegex(path) {
    const paramNames = [];
    const pattern = path
        .replace(/\/+$/, "")  // strip trailing slash
        .replace(/:([A-Za-z_][A-Za-z0-9_]*)/g, (_, name) => {
            paramNames.push(name);
            return "([^/]+)";
        })
        .replace(/\*/g, "(.*)");
    return { regex: new RegExp("^" + pattern + "/?$"), paramNames };
}

export class Router {
    constructor() {
        this._global = [];   // middleware applied to every request
        this._routes = [];   // { method, path, regex, paramNames, handlers }
    }

    use(...handlers) {
        this._global.push(...handlers);
        return this;
    }

    add(method, path, ...handlers) {
        const { regex, paramNames } = pathToRegex(path);
        this._routes.push({ method, path, regex, paramNames, handlers });
        return this;
    }

    async handle(req, ctx = {}) {
        const url = new URL(req.url);
        const path = url.pathname;
        const method = req.method.toUpperCase();
        const enhanced = req;
        enhanced.query = Object.fromEntries(url.searchParams.entries());
        // Run global middleware first; any returning Response short-circuits.
        for (const mw of this._global) {
            const r = await mw(enhanced, ctx);
            if (r instanceof Response) return r;
        }
        // Match route.
        for (const route of this._routes) {
            if (route.method !== method && route.method !== "*") continue;
            const m = route.regex.exec(path);
            if (!m) continue;
            const params = {};
            route.paramNames.forEach((name, i) => { params[name] = m[i + 1]; });
            enhanced.params = params;
            for (const handler of route.handlers) {
                const r = await handler(enhanced, ctx);
                if (r instanceof Response) return r;
            }
            // Route matched but no Response returned → 204.
            return new Response(null, { status: 204 });
        }
        return new Response("Not Found", { status: 404 });
    }

    // Convenience method-bound shortcuts.
    static json(data, init = {}) {
        return new Response(JSON.stringify(data), {
            ...init,
            headers: { "content-type": "application/json", ...(init.headers || {}) },
        });
    }
}

// Decorate Router with .get / .post / .put / .delete / .patch / .options / .head.
for (const m of METHODS) {
    Router.prototype[m.toLowerCase()] = function (path, ...handlers) {
        return this.add(m, path, ...handlers);
    };
}
Router.prototype.all = function (path, ...handlers) {
    return this.add("*", path, ...handlers);
};

export default Router;
