// jwks-verifier/lib/index.mjs
// jwks-verifier — JWKS-based JWT verifier exercising the production
// pattern used by Auth0 / Okta / AWS Cognito / Google Identity.
//
// Real-world security disciplines included:
//   - kid match between token header and JWKS entry (no key reuse)
//   - alg match between token header and JWKS entry's alg claim
//   - use === "sig" check (key not for encryption)
//   - signing-input reconstruction from raw segments (not stringified)
//   - base64url-strict decoding
//   - timing-safe semantics via crypto.subtle (not manual byte compare)

// ── base64url helpers ──────────────────────────────────────────────
function b64uDecodeBytes(s) {
    const b64 = s.replace(/-/g, "+").replace(/_/g, "/") + "=".repeat((4 - s.length % 4) % 4);
    const bin = atob(b64);
    const out = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
    return out;
}
function b64uDecodeStr(s) {
    return new TextDecoder().decode(b64uDecodeBytes(s));
}

// ── algorithm registry ─────────────────────────────────────────────
// Maps JWS alg (RS256/RS384/RS512/PS256/.../ES256/ES384/ES512) to the
// importKey + verify algorithm pair for WebCrypto.
const ALGS = {
    RS256: { import: { name: "RSASSA-PKCS1-v1_5", hash: "SHA-256" }, verify: { name: "RSASSA-PKCS1-v1_5" } },
    RS384: { import: { name: "RSASSA-PKCS1-v1_5", hash: "SHA-384" }, verify: { name: "RSASSA-PKCS1-v1_5" } },
    RS512: { import: { name: "RSASSA-PKCS1-v1_5", hash: "SHA-512" }, verify: { name: "RSASSA-PKCS1-v1_5" } },
    PS256: { import: { name: "RSA-PSS", hash: "SHA-256" }, verify: { name: "RSA-PSS", saltLength: 32 } },
    ES256: { import: { name: "ECDSA", namedCurve: "P-256" }, verify: { name: "ECDSA", hash: "SHA-256" } },
    ES384: { import: { name: "ECDSA", namedCurve: "P-384" }, verify: { name: "ECDSA", hash: "SHA-384" } },
    ES512: { import: { name: "ECDSA", namedCurve: "P-521" }, verify: { name: "ECDSA", hash: "SHA-512" } },
};

// ── public API ─────────────────────────────────────────────────────
export class JwksError extends Error {
    constructor(code, message) { super(message); this.code = code; }
}

export class JwksVerifier {
    constructor(jwks) {
        if (!jwks || !Array.isArray(jwks.keys)) {
            throw new JwksError("invalid_jwks", "JWKS must have a `keys` array");
        }
        this._jwks = jwks;
        this._cache = new Map();   // kid → imported CryptoKey
    }

    async verify(token, expected = {}) {
        if (typeof token !== "string") {
            throw new JwksError("invalid_token", "Token must be a string");
        }
        const parts = token.split(".");
        if (parts.length !== 3) {
            throw new JwksError("invalid_token", "Token must have three dot-separated segments");
        }
        const [hSeg, pSeg, sSeg] = parts;
        let header, payload;
        try {
            header = JSON.parse(b64uDecodeStr(hSeg));
            payload = JSON.parse(b64uDecodeStr(pSeg));
        } catch (_) {
            throw new JwksError("invalid_token", "Header or payload is not valid JSON");
        }
        if (!header.alg) throw new JwksError("missing_alg", "Header missing alg");
        if (!header.kid) throw new JwksError("missing_kid", "Header missing kid");
        const alg = ALGS[header.alg];
        if (!alg) throw new JwksError("unsupported_alg", `Unsupported alg: ${header.alg}`);

        const jwk = this._jwks.keys.find(k => k.kid === header.kid);
        if (!jwk) throw new JwksError("kid_not_found", `kid ${header.kid} not in JWKS`);
        if (jwk.alg && jwk.alg !== header.alg) {
            throw new JwksError("alg_mismatch", "JWKS alg does not match token header alg");
        }
        if (jwk.use && jwk.use !== "sig") {
            throw new JwksError("wrong_use", "JWKS key is not marked for signature use");
        }

        let key = this._cache.get(jwk.kid);
        if (!key) {
            key = await crypto.subtle.importKey("jwk", jwk, alg.import, false, ["verify"]);
            this._cache.set(jwk.kid, key);
        }

        const enc = new TextEncoder();
        const signingInput = enc.encode(hSeg + "." + pSeg);
        const sig = b64uDecodeBytes(sSeg);
        const ok = await crypto.subtle.verify(alg.verify, key, sig, signingInput);
        if (!ok) throw new JwksError("bad_signature", "Signature did not verify");

        // Claim checks (subset of RFC 7519 §4.1).
        if (expected.iss && payload.iss !== expected.iss) {
            throw new JwksError("bad_iss", "iss claim does not match");
        }
        if (expected.aud) {
            const auds = Array.isArray(payload.aud) ? payload.aud : [payload.aud];
            if (!auds.includes(expected.aud)) {
                throw new JwksError("bad_aud", "aud claim does not match");
            }
        }
        if (typeof payload.exp === "number" && expected.now != null && payload.exp < expected.now) {
            throw new JwksError("expired", "Token is expired");
        }
        if (typeof payload.nbf === "number" && expected.now != null && payload.nbf > expected.now) {
            throw new JwksError("not_yet_valid", "Token not yet valid");
        }
        return { header, payload };
    }
}
