// mustache-mini/lib/index.mjs
// mustache-mini — small subset of the Mustache spec
// (https://mustache.github.io/mustache.5.html).
//
// Supported tags:
//   {{name}}        — HTML-escaped interpolation
//   {{{name}}}      — unescaped interpolation
//   {{&name}}       — unescaped interpolation (alias)
//   {{!comment}}    — comments (rendered as empty string)
//   {{#section}}…{{/section}}   — truthy / iteration
//   {{^section}}…{{/section}}   — inverted section
//   {{.}}           — current context (in iteration over primitive arrays)
//   {{a.b.c}}       — dotted-name lookup
//
// Out of scope: partials ({{>name}}), set-delimiter ({{=…=}}), lambdas.

const ESCAPE_MAP = {
    "&": "&amp;",
    "<": "&lt;",
    ">": "&gt;",
    '"': "&quot;",
    "'": "&#x27;",
    "/": "&#x2F;",
    "`": "&#x60;",
    "=": "&#x3D;",
};

function escapeHtml(s) {
    if (s == null) return "";
    const str = String(s);
    return str.replace(/[&<>"'\/`=]/g, (c) => ESCAPE_MAP[c]);
}

// Tokenizer: split template into a flat list of {type, value} tokens.
// type ∈ {"text", "name", "raw", "comment", "open", "close", "inverted"}.
function tokenize(template) {
    const tokens = [];
    const re = /(\{\{\{\s*[\s\S]*?\s*\}\}\})|\{\{([#^/!&>]?)\s*([\s\S]*?)\s*\}\}/g;
    let lastIndex = 0;
    let m;
    while ((m = re.exec(template)) !== null) {
        if (m.index > lastIndex) {
            tokens.push({ type: "text", value: template.slice(lastIndex, m.index) });
        }
        if (m[1]) {
            // {{{name}}} triple-stache
            const inner = m[1].slice(3, -3).trim();
            tokens.push({ type: "raw", value: inner });
        } else {
            const sigil = m[2];
            const name = m[3];
            switch (sigil) {
                case "":  tokens.push({ type: "name", value: name }); break;
                case "&": tokens.push({ type: "raw", value: name }); break;
                case "!": tokens.push({ type: "comment", value: name }); break;
                case "#": tokens.push({ type: "open", value: name }); break;
                case "^": tokens.push({ type: "inverted", value: name }); break;
                case "/": tokens.push({ type: "close", value: name }); break;
                default:  throw new Error("mustache-mini: unsupported sigil " + sigil);
            }
        }
        lastIndex = re.lastIndex;
    }
    if (lastIndex < template.length) {
        tokens.push({ type: "text", value: template.slice(lastIndex) });
    }
    return tokens;
}

// Build a nested AST from the flat token list. Recursion bound by
// well-formedness of #/^ ... / pairs.
function buildAst(tokens, start = 0, endName = null) {
    const nodes = [];
    let i = start;
    while (i < tokens.length) {
        const t = tokens[i];
        if (t.type === "close") {
            if (t.value !== endName) {
                throw new Error(`mustache-mini: mismatched close /${t.value} (expected /${endName})`);
            }
            return { nodes, next: i + 1 };
        }
        if (t.type === "open" || t.type === "inverted") {
            const inner = buildAst(tokens, i + 1, t.value);
            nodes.push({
                type: t.type === "open" ? "section" : "inverted",
                name: t.value,
                children: inner.nodes,
            });
            i = inner.next;
            continue;
        }
        nodes.push(t);
        i++;
    }
    if (endName !== null) {
        throw new Error(`mustache-mini: unclosed section ${endName}`);
    }
    return { nodes, next: i };
}

// Context lookup with dotted names + walk up the context stack.
function lookup(contextStack, name) {
    if (name === ".") return contextStack[contextStack.length - 1];
    const parts = name.split(".");
    for (let i = contextStack.length - 1; i >= 0; i--) {
        let cur = contextStack[i];
        let found = true;
        for (const p of parts) {
            if (cur != null && typeof cur === "object" && p in cur) {
                cur = cur[p];
            } else {
                found = false;
                break;
            }
        }
        if (found) return cur;
    }
    return undefined;
}

function isTruthy(v) {
    if (v == null || v === false || v === 0 || v === "") return false;
    if (Array.isArray(v) && v.length === 0) return false;
    return true;
}

function render(nodes, contextStack) {
    let out = "";
    for (const n of nodes) {
        switch (n.type) {
            case "text":    out += n.value; break;
            case "comment": break;
            case "name": {
                const v = lookup(contextStack, n.value);
                out += escapeHtml(v);
                break;
            }
            case "raw": {
                const v = lookup(contextStack, n.value);
                out += v == null ? "" : String(v);
                break;
            }
            case "section": {
                const v = lookup(contextStack, n.name);
                if (Array.isArray(v)) {
                    for (const item of v) {
                        out += render(n.children, [...contextStack, item]);
                    }
                } else if (isTruthy(v)) {
                    out += render(n.children, [...contextStack, typeof v === "object" ? v : {}]);
                }
                break;
            }
            case "inverted": {
                const v = lookup(contextStack, n.name);
                if (!isTruthy(v)) {
                    out += render(n.children, contextStack);
                }
                break;
            }
            default: throw new Error("mustache-mini: unknown node type " + n.type);
        }
    }
    return out;
}

// Module-level template cache keyed by template string.
const CACHE = new Map();

export function compile(template) {
    let cached = CACHE.get(template);
    if (!cached) {
        const tokens = tokenize(template);
        const ast = buildAst(tokens).nodes;
        cached = (context) => render(ast, [context || {}]);
        CACHE.set(template, cached);
    }
    return cached;
}

export function renderTemplate(template, context) {
    return compile(template)(context);
}

export default { compile, render: renderTemplate };
