// mini-http-server/lib/index.mjs
// mini-http-server — bytes-in → bytes-out HTTP request handler.
//
// Composes globalThis.HTTP (the wire-format codec) with a Bun.serve-style
// fetch-handler API. Real-world consumer: any Bun.serve-shape backend code,
// HTTP-on-fetch servers in Cloudflare Workers / Deno, mock servers in
// integration tests.
//
// Requires globalThis.HTTP to be installed (parseRequest / serializeResponse
// / etc.). The fixture installs a JS reference impl if absent.

function ensureHttp() {
    if (typeof globalThis.HTTP === "undefined" ||
        typeof globalThis.HTTP.parseRequest !== "function") {
        throw new Error("mini-http-server: globalThis.HTTP codec is not installed");
    }
    return globalThis.HTTP;
}

// Build a Web-spec Request from the codec's ParsedRequest shape.
function toRequest(parsed) {
    const HOST = parsed.headers.find(h => h[0].toLowerCase() === "host");
    const host = HOST ? HOST[1] : "localhost";
    const url = "http://" + host + parsed.target;
    const init = {
        method: parsed.method,
        headers: parsed.headers.reduce((h, [n, v]) => (h.append(n, v), h), new Headers()),
    };
    // Body only present for methods that may carry one.
    if (parsed.method !== "GET" && parsed.method !== "HEAD" &&
        parsed.body && parsed.body.length > 0) {
        init.body = parsed.body;
    }
    return new Request(url, init);
}

// Encode a Web-spec Response into raw HTTP/1.1 bytes via the codec.
async function fromResponse(response) {
    const HTTP = ensureHttp();
    const status = response.status;
    const reason = response.statusText || defaultReason(status);
    const headers = [];
    response.headers.forEach((value, name) => { headers.push([name, value]); });
    // Body: read as bytes. Prefer response.bytes() (Uint8Array) over
    // arrayBuffer() since the underlying implementation may return either
    // an ArrayBuffer or a typed-array view depending on the runtime.
    let body;
    if (typeof response.bytes === "function") {
        body = await response.bytes();
    } else {
        const ab = await response.arrayBuffer();
        body = ab instanceof Uint8Array ? ab : new Uint8Array(ab);
    }
    return HTTP.serializeResponse(status, reason, headers, body);
}

const STATUS_REASONS = {
    200: "OK", 201: "Created", 204: "No Content",
    301: "Moved Permanently", 302: "Found", 304: "Not Modified",
    400: "Bad Request", 401: "Unauthorized", 403: "Forbidden", 404: "Not Found",
    409: "Conflict", 422: "Unprocessable Entity",
    500: "Internal Server Error", 501: "Not Implemented", 503: "Service Unavailable",
};
function defaultReason(status) { return STATUS_REASONS[status] || ""; }

// MiniHttpServer({ fetch: async (req) => Response })
//   handle(bytes): bytes — feeds a raw HTTP/1.1 request through the handler
//                          and returns the raw HTTP/1.1 response bytes.
export class MiniHttpServer {
    constructor({ fetch, onError } = {}) {
        if (typeof fetch !== "function") {
            throw new TypeError("mini-http-server: { fetch } handler is required");
        }
        this._fetch = fetch;
        this._onError = onError || ((err) =>
            new Response("Internal Server Error: " + (err.message || String(err)),
                         { status: 500 }));
    }

    async handle(bytes) {
        const HTTP = ensureHttp();
        let parsed, request;
        try {
            parsed = HTTP.parseRequest(bytes);
            request = toRequest(parsed);
        } catch (err) {
            return fromResponse(new Response("Bad Request: " + err.message, { status: 400 }));
        }
        let response;
        try {
            response = await this._fetch(request);
            if (!(response instanceof Response)) {
                throw new TypeError("fetch handler must return a Response");
            }
        } catch (err) {
            response = await this._onError(err, request);
        }
        return fromResponse(response);
    }
}

export default MiniHttpServer;
