// async-pool/lib/index.mjs
// async-pool — concurrent-task pool with AbortSignal cancellation.
//
// Shape mirrors p-queue / p-limit / Promise.all-with-concurrency
// libraries in production. Real-world consumers: fetch-fanout with
// limit, batch upload, parallel test runner, link checker.
//
// Features:
//   - new Pool({ concurrency }): create a pool with N parallel slots
//   - pool.run(task, { signal }): enqueue a task, get a Promise back
//   - signal.aborted at enqueue time → reject without ever running
//   - signal aborted mid-flight → task gets the signal for cooperative
//     bail-out; reject with AbortError once task settles
//   - pool.onIdle(): Promise that resolves when queue + in-flight are empty
//   - pool.abortAll(reason): abort every queued + in-flight task

export class AbortError extends Error {
    constructor(reason) {
        super(typeof reason === "string" ? reason : "Aborted");
        this.name = "AbortError";
    }
}

export class Pool {
    constructor(opts = {}) {
        const { concurrency = 4 } = opts;
        if (!Number.isInteger(concurrency) || concurrency < 1) {
            throw new TypeError("Pool: concurrency must be a positive integer");
        }
        this._concurrency = concurrency;
        this._queue = [];      // pending tasks: {task, signal, resolve, reject}
        this._active = 0;
        this._activeEntries = new Set();  // for abortAll to reject in-flight
        this._idleWaiters = [];
        this._ownSignal = new AbortController();
    }

    get size() { return this._queue.length; }
    get active() { return this._active; }
    get pending() { return this._queue.length + this._active; }

    run(task, opts = {}) {
        const userSignal = opts.signal;
        // Reject immediately if already aborted (spec-correct).
        if (userSignal && userSignal.aborted) {
            return Promise.reject(new AbortError(userSignal.reason));
        }
        return new Promise((resolve, reject) => {
            const entry = { task, userSignal, resolve, reject, cancelled: false };
            // If the user-supplied signal aborts while queued, mark the entry
            // so the dispatcher skips it. The dispatcher reads `cancelled`
            // before running.
            if (userSignal) {
                const onAbort = () => {
                    entry.cancelled = true;
                    entry.reject(new AbortError(userSignal.reason));
                };
                userSignal.addEventListener("abort", onAbort, { once: true });
                entry._abortListener = onAbort;
            }
            this._queue.push(entry);
            this._pump();
        });
    }

    onIdle() {
        if (this._active === 0 && this._queue.length === 0) return Promise.resolve();
        return new Promise((resolve) => this._idleWaiters.push(resolve));
    }

    abortAll(reason) {
        this._ownSignal.abort(reason);
        // Drain queued, never-started tasks.
        const drained = this._queue;
        this._queue = [];
        for (const entry of drained) {
            if (!entry.cancelled) {
                entry.cancelled = true;
                entry.reject(new AbortError(reason));
            }
        }
        // Reject in-flight tasks immediately. Their underlying callbacks
        // may keep running, but the entry.cancelled flag prevents their
        // settled state from leaking back to the consumer.
        for (const entry of this._activeEntries) {
            if (!entry.cancelled) {
                entry.cancelled = true;
                entry.reject(new AbortError(reason));
            }
        }
        this._maybeIdle();
    }

    _pump() {
        while (this._active < this._concurrency && this._queue.length > 0) {
            const entry = this._queue.shift();
            if (entry.cancelled) continue;
            this._active++;
            this._activeEntries.add(entry);
            // Compose a combined signal: aborts if either userSignal OR pool's
            // own abortAll fires.
            const combined = composeSignals(entry.userSignal, this._ownSignal.signal);
            Promise.resolve()
                .then(() => entry.task({ signal: combined.signal }))
                .then((value) => {
                    if (entry._abortListener && entry.userSignal) {
                        entry.userSignal.removeEventListener("abort", entry._abortListener);
                    }
                    combined.dispose();
                    if (!entry.cancelled) entry.resolve(value);
                })
                .catch((err) => {
                    if (entry._abortListener && entry.userSignal) {
                        entry.userSignal.removeEventListener("abort", entry._abortListener);
                    }
                    combined.dispose();
                    if (!entry.cancelled) entry.reject(err);
                })
                .finally(() => {
                    this._active--;
                    this._activeEntries.delete(entry);
                    this._pump();
                    this._maybeIdle();
                });
        }
    }

    _maybeIdle() {
        if (this._active === 0 && this._queue.length === 0) {
            const waiters = this._idleWaiters;
            this._idleWaiters = [];
            for (const w of waiters) w();
        }
    }
}

// Combine N AbortSignals into one that aborts when any source aborts.
// Returns {signal, dispose} where dispose() removes listeners.
function composeSignals(...signals) {
    const real = signals.filter(Boolean);
    if (real.length === 0) {
        return { signal: undefined, dispose: () => {} };
    }
    if (real.length === 1) {
        return { signal: real[0], dispose: () => {} };
    }
    const ctrl = new AbortController();
    const listeners = [];
    for (const s of real) {
        if (s.aborted) {
            ctrl.abort(s.reason);
            break;
        }
        const fn = () => ctrl.abort(s.reason);
        s.addEventListener("abort", fn, { once: true });
        listeners.push([s, fn]);
    }
    return {
        signal: ctrl.signal,
        dispose: () => {
            for (const [s, fn] of listeners) s.removeEventListener("abort", fn);
        },
    };
}

// Convenience: run a batch of N tasks through a pool with concurrency K.
export async function pool(concurrency, tasks, { signal } = {}) {
    const p = new Pool({ concurrency });
    if (signal) {
        if (signal.aborted) throw new AbortError(signal.reason);
        signal.addEventListener("abort", () => p.abortAll(signal.reason), { once: true });
    }
    return Promise.all(tasks.map(t => p.run(t, { signal })));
}

export default { Pool, pool, AbortError };
