// csv-mini/lib/index.mjs
// csv-mini — RFC 4180 CSV parser + writer.
//
// Spec: https://www.rfc-editor.org/rfc/rfc4180
// Implements: comma-separated values; CRLF or LF line endings;
// double-quote field wrapping; "" as escaped quote inside quoted
// fields; embedded commas and newlines inside quoted fields.
//
// Out of scope: BOM stripping, alternate delimiters, header-mode
// row-as-object (provided in a small adapter on top of the
// row-array primitive).

const DEFAULT_OPTS = { delimiter: ",", quote: '"' };

// State machine for parser. Each input character advances a state;
// no regex backtracking, O(n).
const STATE = {
    START_FIELD: 0,
    UNQUOTED: 1,
    QUOTED: 2,
    QUOTED_AFTER_QUOTE: 3,  // inside quoted, just saw a "; might be "" or end-of-field
};

export function parse(input, opts = {}) {
    const { delimiter, quote } = { ...DEFAULT_OPTS, ...opts };
    if (delimiter.length !== 1) throw new Error("csv-mini: delimiter must be 1 char");
    if (quote.length !== 1) throw new Error("csv-mini: quote must be 1 char");
    const rows = [];
    let row = [];
    let field = "";
    let state = STATE.START_FIELD;
    for (let i = 0; i < input.length; i++) {
        const ch = input[i];
        switch (state) {
            case STATE.START_FIELD:
                if (ch === quote) {
                    state = STATE.QUOTED;
                } else if (ch === delimiter) {
                    row.push("");  // empty field
                    // stay in START_FIELD for the next field
                } else if (ch === "\n") {
                    row.push("");
                    rows.push(row);
                    row = [];
                } else if (ch === "\r") {
                    // peek for LF
                    row.push("");
                    rows.push(row);
                    row = [];
                    if (input[i + 1] === "\n") i++;
                } else {
                    field += ch;
                    state = STATE.UNQUOTED;
                }
                break;
            case STATE.UNQUOTED:
                if (ch === delimiter) {
                    row.push(field);
                    field = "";
                    state = STATE.START_FIELD;
                } else if (ch === "\n") {
                    row.push(field);
                    field = "";
                    rows.push(row);
                    row = [];
                    state = STATE.START_FIELD;
                } else if (ch === "\r") {
                    row.push(field);
                    field = "";
                    rows.push(row);
                    row = [];
                    if (input[i + 1] === "\n") i++;
                    state = STATE.START_FIELD;
                } else {
                    field += ch;
                }
                break;
            case STATE.QUOTED:
                if (ch === quote) {
                    state = STATE.QUOTED_AFTER_QUOTE;
                } else {
                    field += ch;
                }
                break;
            case STATE.QUOTED_AFTER_QUOTE:
                if (ch === quote) {
                    field += quote;  // escaped quote
                    state = STATE.QUOTED;
                } else if (ch === delimiter) {
                    row.push(field);
                    field = "";
                    state = STATE.START_FIELD;
                } else if (ch === "\n") {
                    row.push(field);
                    field = "";
                    rows.push(row);
                    row = [];
                    state = STATE.START_FIELD;
                } else if (ch === "\r") {
                    row.push(field);
                    field = "";
                    rows.push(row);
                    row = [];
                    if (input[i + 1] === "\n") i++;
                    state = STATE.START_FIELD;
                } else {
                    throw new Error(`csv-mini: malformed quoted field at offset ${i} (char ${JSON.stringify(ch)})`);
                }
                break;
        }
    }
    // Flush trailing field/row.
    if (state === STATE.UNQUOTED || state === STATE.QUOTED_AFTER_QUOTE) {
        row.push(field);
        rows.push(row);
    } else if (state === STATE.START_FIELD && row.length > 0) {
        row.push("");
        rows.push(row);
    } else if (state === STATE.QUOTED) {
        throw new Error("csv-mini: unterminated quoted field");
    }
    return rows;
}

const NEEDS_QUOTING = /[,"\r\n]/;

export function stringify(rows, opts = {}) {
    const { delimiter, quote } = { ...DEFAULT_OPTS, ...opts };
    const escapeRe = new RegExp(`[${delimiter}${quote}\r\n]`);
    const quoteRe = new RegExp(quote, "g");
    const lines = [];
    for (const row of rows) {
        const fields = row.map((f) => {
            const s = f == null ? "" : String(f);
            if (escapeRe.test(s)) {
                return quote + s.replace(quoteRe, quote + quote) + quote;
            }
            return s;
        });
        lines.push(fields.join(delimiter));
    }
    return lines.join("\r\n");
}

// header-mode row-as-object adapter (production-shape).
export function parseObjects(input, opts = {}) {
    const rows = parse(input, opts);
    if (rows.length === 0) return [];
    const [header, ...data] = rows;
    return data.map((row) => {
        const obj = {};
        for (let i = 0; i < header.length; i++) obj[header[i]] = row[i] ?? "";
        return obj;
    });
}

export default { parse, stringify, parseObjects };
