// signals-mini/lib/index.mjs
// signals-mini — small reactive-signals library.
//
// Shape mirrors Preact signals / Solid signals / Vue 3 ref+computed.
// Real-world consumers: fine-grained reactivity in web UIs, runtime
// state graphs, computed-derived state.
//
// Primitives:
//   signal(initial)          → {get value, set value, peek()}
//   computed(fn)             → {get value, peek()} — lazy + cached
//   effect(fn) → unsubscribe → reruns whenever a tracked dep changes
//   batch(fn)                → defers effect re-runs until fn returns
//   untracked(fn)            → reads values without registering deps
//
// Implementation notes:
//   - Subscriber tracking via a single CURRENT_OBSERVER variable.
//   - Diamond dependencies resolved by per-effect "dirty" flag plus
//     topological-ish dispatch: each affected node is marked once.
//   - Manual cleanup (no FinalizationRegistry available in basin).

let CURRENT_OBSERVER = null;
let BATCH_DEPTH = 0;
let PENDING_EFFECTS = new Set();

function notify(subscribers) {
    // Implicit batching: every notify() opens an implicit batch so that
    // diamond dependencies (a → b, a → c, b+c → d) settle b and c BEFORE
    // d gets re-run, firing d only once. The outermost notify drains.
    const outermost = BATCH_DEPTH === 0;
    BATCH_DEPTH++;
    try {
        for (const s of subscribers) PENDING_EFFECTS.add(s);
    } finally {
        BATCH_DEPTH--;
        if (outermost && BATCH_DEPTH === 0) drain();
    }
}

function drain() {
    // Dedup across the whole drain pass — an effect re-scheduled while
    // draining (because a computed it depends on marks dirty later) is
    // only fired once. Subscribers added during run() go to PENDING_EFFECTS
    // and are drained in subsequent iterations.
    const fired = new Set();
    BATCH_DEPTH++;  // keep nested notifies deferred during drain
    try {
        while (PENDING_EFFECTS.size > 0) {
            const batch_ = PENDING_EFFECTS;
            PENDING_EFFECTS = new Set();
            for (const o of batch_) {
                if (fired.has(o) || o._disposed) continue;
                fired.add(o);
                o._run();
            }
        }
    } finally {
        BATCH_DEPTH--;
    }
}

export function signal(initial) {
    let value = initial;
    const subscribers = new Set();
    return {
        get value() {
            if (CURRENT_OBSERVER) {
                subscribers.add(CURRENT_OBSERVER);
                CURRENT_OBSERVER._deps.add(subscribers);
            }
            return value;
        },
        set value(v) {
            if (Object.is(v, value)) return;
            value = v;
            notify(subscribers);
        },
        peek() { return value; },
    };
}

export function computed(fn) {
    let cached;
    let dirty = true;
    const subscribers = new Set();
    const ownObserver = {
        _deps: new Set(),
        _run() {
            // When a dep changes, mark dirty and notify our own subscribers.
            // We don't recompute until someone reads us — that's the
            // "computed is lazy" guarantee.
            if (!dirty) {
                dirty = true;
                notify(subscribers);
            }
        },
    };
    function recompute() {
        // Remove old subscriptions.
        for (const subs of ownObserver._deps) subs.delete(ownObserver);
        ownObserver._deps.clear();
        // Re-run with our observer current.
        const prev = CURRENT_OBSERVER;
        CURRENT_OBSERVER = ownObserver;
        try { cached = fn(); }
        finally { CURRENT_OBSERVER = prev; }
        dirty = false;
    }
    return {
        get value() {
            if (dirty) recompute();
            if (CURRENT_OBSERVER) {
                subscribers.add(CURRENT_OBSERVER);
                CURRENT_OBSERVER._deps.add(subscribers);
            }
            return cached;
        },
        peek() {
            if (dirty) recompute();
            return cached;
        },
    };
}

export function effect(fn) {
    const observer = {
        _deps: new Set(),
        _disposed: false,
        _run() {
            if (observer._disposed) return;
            // Unsubscribe from previous deps.
            for (const subs of observer._deps) subs.delete(observer);
            observer._deps.clear();
            const prev = CURRENT_OBSERVER;
            CURRENT_OBSERVER = observer;
            try { fn(); }
            finally { CURRENT_OBSERVER = prev; }
        },
    };
    observer._run();
    return function dispose() {
        observer._disposed = true;
        for (const subs of observer._deps) subs.delete(observer);
        observer._deps.clear();
    };
}

export function batch(fn) {
    BATCH_DEPTH++;
    try {
        return fn();
    } finally {
        BATCH_DEPTH--;
        if (BATCH_DEPTH === 0) drain();
    }
}

export function untracked(fn) {
    const prev = CURRENT_OBSERVER;
    CURRENT_OBSERVER = null;
    try { return fn(); }
    finally { CURRENT_OBSERVER = prev; }
}

export default { signal, computed, effect, batch, untracked };
